File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if nums[0] < nums[-1]:
            return nums[0]
        l = 0
        res = nums[0]
        r = len(nums) - 1
        while l <= r:
            if nums[l] < nums[r]:
                res = min(res, nums[l])
            mid = (l + r) // 2
            res = min(res, nums[mid])
            if nums[l] <= nums[mid]:
                l = mid + 1
            else:
                r = mid - 1
        return res

File: Data Structures & Algorithms/eating-bananas/submission-0.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int st = 1;
        int dr = 0;
        int n = piles.size();
        for (int i = 0; i < n; i++) {
            if (piles[i] > dr) dr = piles[i];
        }
        int res = dr;
        while (st <= dr) {
            int mid = st + (dr - st) / 2;
            long long hours = 0;
            for (int i = 0; i < n; i++) {
                hours += (piles[i] + mid - 1) / mid;
            }
            if (hours <= h) res = mid, dr = mid - 1;
            else st = mid + 1;
        }
        return res;
    }
};
